File: modules/stackr-card-vision/ios/StackrCardVisionModule.swift
import CoreImage
import ExpoModulesCore
import Foundation
import UIKit

private let stackrCardVisionModuleVersion = "stackr-card-vision-native-v1"

public class StackrCardVisionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("StackrCardVision")
    Events(stackrCardFrameAnalysisEvent)

    OnCreate {
      StackrCardFrameAnalyserInstrumentation.attach(self)
    }

    OnDestroy {
      StackrCardFrameAnalyserInstrumentation.detach(self)
    }

    Function("getCardVisionRuntimeInfo") { () -> [String: Any?] in
      let onnxBridgeClassAvailable = classAvailable("OnnxruntimeModule")
      let frameClassAvailable = classAvailable("Frame")
      let frameProcessorClassAvailable = classAvailable("FrameProcessor") || classAvailable("FrameProcessorPlugin")
      let workletsCoreClassAvailable = classAvailable("WorkletsModule")
      _ = CIContext()
      let ciContextAvailable = true

      return [
        "platform": "ios",
        "moduleVersion": stackrCardVisionModuleVersion,
        "onnxRuntimeAvailable": onnxBridgeClassAvailable,
        "cameraFrameAccessAvailable": frameClassAvailable && frameProcessorClassAvailable && workletsCoreClassAvailable,
        "nativeImageProcessingAvailable": ciContextAvailable,
        "onnxRuntimeDetail": "bridgeClass=\(onnxBridgeClassAvailable)",
        "cameraFrameAccessDetail": "visionCameraFrame=\(frameClassAvailable); frameProcessor=\(frameProcessorClassAvailable); workletsCore=\(workletsCoreClassAvailable); pluginRegistered=false",
        "nativeImageProcessingDetail": "coreImage=\(ciContextAvailable); luminanceAnalyser=\(CardFrameAnalyserQualityConfig.default.version); rectifier=stackr-card-rectification-v1.0.0; roi=stackr-pokemon-card-roi-v1.0.0; system=\(UIDevice.current.systemName) \(UIDevice.current.systemVersion)",
        "opencvAvailable": false,
        "opencvVersion": nil
      ]
    }

    Function("runCardFrameAnalyserFixtureTests") { () -> [String: Any?] in
      StackrCardFrameAnalyserFixtures.runFixtureTests()
    }

    Function("benchmarkCardFrameAnalyserFixtures") { (fixtureCount: Int?) -> [String: Any?] in
      StackrCardFrameAnalyserFixtures.benchmark(fixtureCount: fixtureCount ?? 120)
    }

    Function("rectifyCapturedCard") { (request: [String: Any?]) -> [String: Any?] in
      StackrCardRectifier.rectify(request)
    }

    Function("deleteCardRectificationOutputs") { (scanId: String) -> [String: Any?] in
      StackrCardRectifier.deleteOutputs(scanId: scanId)
    }

    Function("loadCardIdentitySearchCatalogue") { (request: [String: Any?]) -> [String: Any?] in
      StackrCardIdentitySearchEngine.load(request)
    }

    Function("searchCardIdentityEmbedding") { (request: [String: Any?]) -> [String: Any?] in
      StackrCardIdentitySearchEngine.search(request)
    }

    Function("benchmarkCardIdentitySearch") { (request: [String: Any?]?) -> [String: Any?] in
      StackrCardIdentitySearchEngine.benchmark(request ?? [:])
    }

    Function("resetCardIdentitySearchCatalogue") { () -> [String: Any?] in
      StackrCardIdentitySearchEngine.reset()
    }

    Function("getCardFrameAnalyserInstrumentation") { () -> [String: Any?] in
      StackrCardFrameAnalyserInstrumentation.snapshot()
    }

    Function("resetCardFrameAnalyserInstrumentation") { () -> [String: Any?] in
      StackrCardFrameAnalyserInstrumentation.reset()
      return StackrCardFrameAnalyserInstrumentation.snapshot()
    }

    Function("recordCardFrameAnalyserFocusFailure") { () -> [String: Any?] in
      StackrCardFrameAnalyserInstrumentation.markFocusFailure()
      return StackrCardFrameAnalyserInstrumentation.snapshot()
    }
  }
}

private func classAvailable(_ className: String) -> Bool {
  return NSClassFromString(className) != nil
}

File: modules/stackr-card-vision/android/src/main/java/com/stackr/cardvision/StackrCardVisionModule.kt
package com.stackr.cardvision

import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private const val STACKR_CARD_VISION_MODULE_VERSION = "stackr-card-vision-native-v1"

class StackrCardVisionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("StackrCardVision")
    Events(STACKR_CARD_FRAME_ANALYSIS_EVENT)

    OnCreate {
      StackrCardFrameAnalyserInstrumentation.attach(this@StackrCardVisionModule)
      try {
        StackrCardFrameAnalyserPluginRegistration.register()
      } catch (_: Throwable) {
        // Diagnostics report the unavailable frame processor runtime below.
      }
    }

    OnDestroy {
      StackrCardFrameAnalyserInstrumentation.detach(this@StackrCardVisionModule)
    }

    Function("getCardVisionRuntimeInfo") {
      val onnxBridgeClassAvailable = classAvailable("ai.onnxruntime.reactnative.OnnxruntimeModule")
      val onnxJsiLibraryLoadable = nativeLibraryLoadable("onnxruntimejsi")
      val frameClassAvailable = classAvailable("com.mrousavy.camera.frameprocessors.Frame")
      val framePluginClassAvailable = classAvailable("com.mrousavy.camera.frameprocessors.FrameProcessorPlugin")
      val workletsCoreClassAvailable = classAvailable("com.margelo.worklets.Worklets")
      val workletsClassAvailable = classAvailable("com.swmansion.worklets.WorkletsModule")
      val bitmapFactoryAvailable = classAvailable("android.graphics.BitmapFactory")

      mapOf(
        "platform" to "android",
        "moduleVersion" to STACKR_CARD_VISION_MODULE_VERSION,
        "onnxRuntimeAvailable" to (onnxBridgeClassAvailable && onnxJsiLibraryLoadable),
        "cameraFrameAccessAvailable" to (
          frameClassAvailable &&
            framePluginClassAvailable &&
            StackrCardFrameAnalyserPluginRegistration.registered &&
            workletsCoreClassAvailable
          ),
        "nativeImageProcessingAvailable" to bitmapFactoryAvailable,
        "onnxRuntimeDetail" to "bridgeClass=$onnxBridgeClassAvailable; jsiLibraryLoadable=$onnxJsiLibraryLoadable",
        "cameraFrameAccessDetail" to "visionCameraFrame=$frameClassAvailable; frameProcessorPlugin=$framePluginClassAvailable; pluginRegistered=${StackrCardFrameAnalyserPluginRegistration.registered}; reactNativeWorkletsCore=$workletsCoreClassAvailable; reactNativeWorklets=$workletsClassAvailable",
        "nativeImageProcessingDetail" to "androidBitmapFactory=$bitmapFactoryAvailable; luminanceAnalyser=${CardFrameAnalyserQualityConfig.default.version}; rectifier=stackr-card-rectification-v1.0.0; roi=stackr-pokemon-card-roi-v1.0.0; sdk=${Build.VERSION.SDK_INT}",
        "opencvAvailable" to false,
        "opencvVersion" to null
      )
    }

    Function("runCardFrameAnalyserFixtureTests") {
      StackrCardFrameAnalyserFixtures.runFixtureTests()
    }

    Function("benchmarkCardFrameAnalyserFixtures") { fixtureCount: Int? ->
      StackrCardFrameAnalyserFixtures.benchmark(fixtureCount ?: 120)
    }

    Function("rectifyCapturedCard") { request: Map<String, Any?> ->
      StackrCardRectifier.rectify(request, appContext.reactContext?.cacheDir)
    }

    Function("deleteCardRectificationOutputs") { scanId: String ->
      StackrCardRectifier.deleteOutputs(scanId, appContext.reactContext?.cacheDir)
    }

    Function("loadCardIdentitySearchCatalogue") { request: Map<String, Any?> ->
      StackrCardIdentitySearchEngine.load(request)
    }

    Function("searchCardIdentityEmbedding") { request: Map<String, Any?> ->
      StackrCardIdentitySearchEngine.search(request)
    }

    Function("benchmarkCardIdentitySearch") { request: Map<String, Any?>? ->
      StackrCardIdentitySearchEngine.benchmark(request ?: emptyMap())
    }

    Function("resetCardIdentitySearchCatalogue") {
      StackrCardIdentitySearchEngine.reset()
    }

    Function("getCardFrameAnalyserInstrumentation") {
      StackrCardFrameAnalyserInstrumentation.snapshot()
    }

    Function("resetCardFrameAnalyserInstrumentation") {
      StackrCardFrameAnalyserInstrumentation.reset()
      StackrCardFrameAnalyserInstrumentation.snapshot()
    }

    Function("recordCardFrameAnalyserFocusFailure") {
      StackrCardFrameAnalyserInstrumentation.markFocusFailure()
      StackrCardFrameAnalyserInstrumentation.snapshot()
    }
  }

  private fun classAvailable(className: String): Boolean {
    return try {
      Class.forName(className)
      true
    } catch (_: Throwable) {
      false
    }
  }

  private fun nativeLibraryLoadable(libraryName: String): Boolean {
    return try {
      System.loadLibrary(libraryName)
      true
    } catch (_: Throwable) {
      false
    }
  }
}
